### ISSUE TITLE
Files tags-xx  and files with translated help files should not be in the runtime/doc directory

### ISSUE THREAD

**Issue by azc100** [vim/vim#16061]
### Steps to reproduce

root ~/vim/runtime/doc > ls -l tags*
-rw-r--r-- 1 root root 400313 nov 15 09:16 tags
-rw-r--r-- 1 root root  41035 ago 12 08:35 tagsrch.txt
-rw-r--r-- 1 root root    437 nov 15 09:20 tags-ru
root ~/vim/runtime/doc >

### Expected behaviour

root ~/vim/runtime/doc > ls -l tags*
-rw-r--r-- 1 root root 400313 nov 15 09:16 tags
-rw-r--r-- 1 root root  41035 ago 12 08:35 tagsrch.txt
root ~/vim/runtime/doc >

Said in other words, as far as I can tell, translations of help files in other languages are kept in external sites, not in in the runtime/doc Vim distribution directory. And the "national" tag files tags-xx should not be there either. 

E.g., the available Italian translations of the 9.1 Vim help files (including the file tags-it) can be found here:
https://sites.google.com/view/vimdoc-it/vim-9-1

If this is going to change, then a lot more files should go to the runtime/doc directory, and its maintenance should get much more complex.

### Version of Vim

VIM - Vi IMproved 9.1.866

### Environment

Linux Mint 22 Wilma
xterm-256color
/bin/bash

### Logs and stack traces

```shell
N/A
```

**Comment by chrisbra** [comment]
We had those discussions over mail and I asked to discuss it more publicly. But now that I check it again, it seems those might be created automatically? We don't have those currently in the repository, if I see that correctly.

**Comment by azc100** [comment]
Hi Chris,
I had to login into the Vim git page to create the issue, it did not happen
by magic. I am happy if some other developer has his/her say.

All the best, Antonio

Il ven 15 nov 2024, 12:29 Christian Brabandt ***@***.***> ha
scritto:

**Comment by chrisbra** [comment]
Well, the `tags-*` file seems to be ignored in the ignorefiles, you can check using `git status --ignored`.
Also, if you browse github directory, you won't see the `tags-ru` file.

It seems to be generated by `make vimtags` in `runtime/doc` directory, but apparently never cleaned up.

So perhaps we just need this rule here:
```diff
diff --git a/runtime/doc/Makefile b/runtime/doc/Makefile
index 785128e1f..2b9260933 100644
--- a/runtime/doc/Makefile
+++ b/runtime/doc/Makefile
@@ -93,7 +93,7 @@ test_urls:
        $(VIMPROG) --clean -S test_urls.vim

 clean:
-       -rm -f doctags *.html tags.ref
+       -rm -f doctags *.html tags.ref tags-*

 # These files are in the extra archive, skip if not present

```
to clean up properly?

**Comment by RestorerZ** [comment]
Actually, this is not a problem, but the expected behavior of Vim. And this has already been discussed several [times](https://github.com/vim/vim/pull/14479#issuecomment-2053995559).
One of the options to change this, I once suggested [here](https://github.com/vim/vim/pull/14571).

**Comment by chrisbra** [comment]
I never use `make vimtags`. I usually just use `make tags`. When does this happen?

**Comment by RestorerZ** [comment]
Tag creation can be either by installing Vim, or by executing the “doc/make Makefile” file.
Here is the code from “src/Makefile”
```makefile
# Generate the help tags with ":helptags" to handle all languages.
# Move the distributed tags file aside and restore it, to avoid it being
# different from the repository.
	cd $(HELPSOURCE); if test -z "$(CROSS_COMPILING)" -a -f tags; then \
		mv -f tags tags.dist; fi
	@echo generating help tags
	-@BUILD_DIR=`pwd`; cd $(HELPSOURCE); if test -z "$(CROSS_COMPILING)"; then \
		$(MAKE) VIMPROG="$$BUILD_DIR/$(VIMTARGET)" vimtags; fi
	cd $(HELPSOURCE); \
		files=`ls *.txt tags`; \
		files="$$files `ls *.??x tags-?? 2>/dev/null || true`"; \
		$(INSTALL_DATA) $$files  $(DEST_HELP); \
		cd $(DEST_HELP); \
		chmod $(HELPMOD) $$files
	$(INSTALL_DATA)  $(HELPSOURCE)/*.pl $(DEST_HELP)
	chmod $(SCRIPTMOD) $(DEST_HELP)/*.pl
	cd $(HELPSOURCE); if test -f tags.dist; then mv -f tags.dist tags; fi
```

**Comment by RestorerZ** [comment]
If, for example, @azc100 wants the Italian license agreement to be displayed
when the Vim self-installer is created, it requires placing the "uganda.itx" in
the “doc” directory of the Vim repository. And once this is done, “tags-it”
files will also be created.

Is there another way to have the national language license agreement added when
the installer is created, but not require the “uganda.??x” file to be placed in
the “runtime/doc” directory.
However, for this purpose it is necessary to change a bit the files
“doc/Makefile” ("doc/Make_mvc.mak") and “nsis/gvim.nsi”.

**Comment by azc100** [comment]
Chris, "RestorerZ",

 clean:
-       -rm -f doctags *.html tags.ref
+       -rm -f doctags *.html tags.ref tags-*

As a matter of fact (after getting a fresh copy of Vim, with "git clone"), tags-ru is not present in the distribution, but it appears after invoking "make install". But the other translated file "uganda.rux" IS in the distribution (and I can admire it, inside my Italian flavored Vim, typing ":h uganda@ru").

So, to make the "rm" complete, the modified "rm" line above should be:

+       -rm -f doctags *.html tags.ref tags-* *.??x

By the way, the fact that "tags-ru" is introduced by the "make install" process has this (probably unintended) consequence:

root /usr/local/share/vim/vim91/doc > ls -l uganda*  tags*
-rw-r--r-- 1 root  root  400357 nov 16 08:56 tags
-rw-r--r-- 1 root  root  102669 nov 15 09:52 tags-it
-rw-r--r-- 1 root  root   41035 nov 16 08:56 tagsrch.txt
-rw-r--r-- 1 root  root     437 nov 16 08:56 tags-ru
-rw-r--r-- 1 azc10 azc10  15786 ago 15 09:55 uganda.itx
-rw-r--r-- 1 root  root   29670 nov 16 08:56 uganda.rux
-rw-r--r-- 1 root  root   12933 nov 16 08:56 uganda.txt
root /usr/local/share/vim/vim91/doc > 

Said in other terms, in my Italian $VIMRUNTIME/doc directory I find:
tags-it / uganda.itx
(I put them there, together with the Italian version of the Vim documentations, in a way that is absolutely independent from the Vim distribution)
AND
tags-ru / uganda.rux
that in my humble opinion should not be there at all.

Note the different size of the "tags-it" file compared with the "tags-ru" file. The real reason for having a "tags-ll" file is to be able to display the help in a given national language, not just to be able to display the "uganda.llx" file.

But all of the above does not address the worries of "restorer":

> If, for example, @azc100 wants the Italian license agreement to be displayed
> when the Vim self-installer is created, it requires placing the "uganda.itx" in
> the “doc” directory of the Vim repository. And once this is done, “tags-it”
> files will also be created.

This will only augment the pollution of uganda.llx and tags-ll files, that will end up in EACH installation (as they ended up in my installation). In my humble opinion the only license to be used is the English one (no matter what the language in use is) and the same applies to "uganda.txt"

In the same line (and for the same reason and with the same problems), in the distribution we have:

root ~/vim > ls -ld README*
drwxr-xr-x 2 root root  4096 nov 15 23:00 READMEdir
-rw-r--r-- 1 root root  6914 nov 15 23:00 README.md
-rw-r--r-- 1 root root 11498 nov 15 23:00 README.rux.txt
-rw-r--r-- 1 root root  5081 nov 15 23:00 README.txt
-rw-r--r-- 1 root root 10833 nov 15 23:00 README_VIM9.md

...where README.rux.txt should not be there. Or, we will finish up having a proliferation of README.llx files as well.

> Is there another way to have the national language license agreement added when
> the installer is created, but not require the “uganda.??x” file to be placed in
> the “runtime/doc” directory.

I would suggest to include both files (README.rux.txt and uganda.rux) in the installer itself.  They don't change often.

> However, for this purpose it is necessary to change a bit the files
> “doc/Makefile” ("doc/Make_mvc.mak") and “nsis/gvim.nsi”.

I feel that all the above should not belong to any of the above Makefiles. Said in other terms, all the ".rux" files should go away from the Vim distribution (as should all the possible ".llx" files, I have nothing against the Russian language in itself).

To really change the "English only" README/uganda matter, a decision should be taken to INCLUDE the translations of the Vim documentation in other languages, but this will very likely create more problems than it solves.

Hope it helps, Antonio

**Comment by RestorerZ** [comment]
@azc100 

> In my humble opinion the only license to be used is the English one (no matter what the language in use is) and the same applies to "uganda.txt"

Antonio, as I understand, you are fluent in English and can easily read both the
license agreement and the “readme” file. You also assume that other users will
be able to read it as well. In this case the question arises - why translate the
same documentation if the user can easily read English?
I assume that not all users of the Vim editor are so fluent in English. And that
reading both the license agreement and the “redme” file in the user's native
language will be more pleasant and better.

> But the other translated file "uganda.rux" IS in the distribution (and I can admire it, inside my Italian flavored Vim, typing ":h uganda@ru").

That an Italian user or any other does not need documentation in a foreign
language in their installed Vim editor, I completely agree with you and share
that point of view.
That's why for the MS Windows installer I tried to provide for this and [exclude](https://github.com/vim/vim/blob/5c66e23c624717216d380d938d0bba5d34a004fe/nsis/gvim.nsi#L624-L727)
such files from installation on the user's machine.
I was [planning](https://github.com/RestorerZ/fork_vim-orig/blob/ed7d847fd5410c51f127b2e1db3a7b29e6322f1c/src/Makefile#L2358-L2359) to do something similar for the installer in UNIX-like systems,
but I haven't had time to do it yet. Besides, I am not that familiar with the
UNIX-like world.

So, this can all be solved with a few lines of code without depriving the user
of the ability to interact with Vim in their native language and without
clogging up their system with unnecessary files.

**Comment by chrisbra** [comment]
I am sorry, I don't follow quite what problem we are solving. I can change the Makefile as suggested, but I am missing what exact changes should be done. But I don't think that the distribution of translated tag files / READMEs is an actual issue. What if the user wants to read translated help files / READMEs?

**Comment by azc100** [comment]
At @RestorerZ:

> > In my humble opinion the only license to be used is the English
> > one (no matter what the language in use is) and the same applies
> > to "uganda.txt"

> Antonio, as I understand, you are fluent in English and can easily
> read both the license agreement and the “readme” file.

Correct! 

> You also assume that other users will be able to read it as well.

Correct, but with a twist. Users are not really interested to the
license. If they want to install Vim, accepting the license is
mandatory, no matter what the license actually says. So, it does not
really matter the language in which the license is written, they have
to accept it anyway.

> In this case the question arises - why translate the same
> documentation if the user can easily read English? I assume that not
> all users of the Vim editor are so fluent in English. And that
> reading both the license agreement and the “readme” file in the
> user's native language will be more pleasant and better.

This too is correct!

So, my proposal is: let us solve it the way the "menu" problem was
solved.

In practice this would mean: Let us add, in the same directory in
which the README.txt file is ( vim/runtime in the distribution ) , a
second file named LICENSE.txt (that contains almost the same stuff as
the uganda.txt files, but without the pointers etc.). I can prepare
it, and let Chris have it.

In the same vim/runtime directory there already is a "lang" directory,
with all the translated menus. Here I would add a README.ru and a
LICENSE.ru texts (independent from the "doc" directory), and I would
gladly prepare a README.it and a LICENSE.it files, just for testing
purpose. Then a message to the group of the Vim translators could be
sent to further populate the "lang" directory with the files README.ll
and LICENSE.LL in other languages...
The list of the maintainers can easily be built using "grep" against
the files in the above mentioned vim/runtime/lang directory.

The logic (during the installation) would be: IF in the
vim/runtime/lang directory there is a README.ll and a LICENSE.ll for
language "ll", then use them. If not use the original English files.

> > But the other translated file "uganda.rux" IS in the distribution
> > (and I can admire it, inside my Italian flavored Vim, typing ":h
> > uganda@ru").

This is correct as well. The solution proposed above would avoid this.

> That an Italian user or any other does not need documentation in a
> foreign language in their installed Vim editor, I completely agree
> with you and share that point of view.
> That's why for the MS Windows installer I tried to provide for this
> and exclude such files from installation on the user's machine. I
> was planning to do something similar for the installer in UNIX-like
> systems, but I haven't had time to do it yet. Besides, I am not that
> familiar with the UNIX-like world.

If my proposal above will be accepted, then this problem is solved. As
it is now (in Linux), I get the Russian uganda.rux file (and the
corresponding tags-ru file) no matter what. I have nothing against
the files themselves, only I don't need them.


At Chris:

> I am sorry, I don't follow quite what problem we are solving. I can
> change the Makefile as suggested, but I am missing what exact
> changes should be done.

I hope you can agree with the above proposed solution.
The interested Makefile can stay as it is now.

> But I don't think that the distribution of translated tag files /
> READMEs is an actual issue. What if the user wants to read
> translated help files / READMEs?

During the installation, only the README and the LICENSE files are
really needed. They are (or will be) in the vim/runtime or in the
vim/runtime/lang directory. Not interferences with the vim/runtime/doc
directory.

If the user wants help in his/her own language, he/she can get it from
the page:
https://www.vim.org/translations.php

Hope it helps, Antonio

**Comment by chrisbra** [comment]
Sure, can you please create a patch/PR for this?

**Comment by RestorerZ** [comment]
@azc100 
Antonio, how will these files be installed on the system? On what signs and where will the condition for the command to install the localized license and readmе files be checked.

Also please note that if we use the “License.*” file instead of the “uganda.*” file, we will need to make changes in the translation files of the installer for MS Windows. The files “doc/Makefile”, “doc/Make_mvc.mak” and probably a number of other files. In principle, I could do it, but not until the end of this month.
And among other things, the “License” file from the root of the repository is used to display the license agreement on Github.

But, in principle, the idea of using the “license.*” file instead of “uganda.*” to display the license agreement in the installer is good in itself. As well as a separate directory for translated readme and license files.

**Comment by azc100** [comment]
> Antonio, how will these files be installed on
> the system? On what signs and where will the
> condition for the command to install the
> localized license and readmе files be checked.

The LICENSE.txt file should be in the "vim/"
directory, where his "friend", the README.txt file
already lives. So, if the installer finds the
README.txt file it will also find the LICENSE.txt
file.

> Also please note that if we use the “License.*”
> file instead of the “uganda.*” file, we will
> need to make changes in the translation files of
> the installer for MS Windows. The files
> “doc/Makefile”, “doc/Make_mvc.mak” and probably
> a number of other files.
> In principle, I could do it, but not until the
> end of this month.

Time should not be a problem. Up to then there
just will be a few extra files, not used
(the LICENSE.txt file and  the lang/* files.

The logic of the installer should be changed
to check if there are "national" files in the
vim/lang directory and, if this is the case,
use README.ll and LICENSE.ll instead of
README.txt and LICENSE.txt.
The README.rux file should (in future) be
renamed README.ru and moved to the vim/lang
directory. And the LICENSE.ru file, built
from the runtime/doc/uganda.rux should go
to the vim/lang directory.

> And among other things, the “License” file from
> the root of the repository is used to display
> the license agreement on Github.

If this is the case, the LICENSE.txt file (which
I prepared using the runtime/doc/uganda.txt file)
could be shortened. It is just a subset of what
in in "uganda.txt". Of course the "uganda.txt"
file will remain as it is in the runtime/doc
directory.

> But, in principle, the idea of using the
> “license.*” file instead of “uganda.*” to
> display the license agreement in the installer
> is good in itself. As well as a separate
> directory for translated readme and license
> files.

Very good! Let us make it work!

Antonio

**Comment by RestorerZ** [comment]
That the logic of the installer should be changed is understandable. The question is how it should be changed.
If on MS Windows system I can somehow imagine the process of checking the condition of what language is used (and it is already implemented in the installer), then on UNIX-like systems I don't understand how to determine the language. By environment variable $LANG? By LC_MESSAGE? Also, as I recall, UNIX-like systems use both the “src/Makefile” and the “src/install-sh” command file to install the Vim editor

**Comment by azc100** [comment]
I can't help with the Windows Installer, but I can help with the Unix one. I would use the first two bytes of $LANG, or the $LANGUAGE variable (if present) to find out what the local language is. I usually just install Vim with "make install", but I can look into the shell scripts (I am much less familiar with Makefiles).

**Comment by RestorerZ** [comment]
@azc100 
Yes, Antonio, I was thinking about those variables too. The nuance is that UINIX-like systems, they are only similar and each has its own differences. The same Linux distributions can be incompatible with each other, I'm not talking about BSD-like and MacOS. But the basis is the same, so everything is solvable. Practically anything can be done, just need time to do it. I've already started to read the documentation on this case.

As for the MS Windows installer, it's practically done there
https://github.com/vim/vim/blob/9a65a8c270f3f101b59b63f5252c7e96549bed2d/nsis/gvim.nsi#L641C1-L655C7
```nsis
!if /FileExists ..\README.itx.txt
    ${If} $Language = ${LANG_ITALIAN}
	File ..\README.itx.txt
    ${EndIf}
!endif
!if /FileExists ..\README.jax.txt
    ${If} $Language = ${LANG_JAPANESE}
	File ..\README.jax.txt
    ${EndIf}
!endif
!if /FileExists ..\README.rux.txt
    ${If} $Language = ${LANG_RUSSIAN}
	File ..\README.rux.txt
    ${EndIf}
!endif
```
, just need to change the file names and some other things
https://github.com/vim/vim/blob/9a65a8c270f3f101b59b63f5252c7e96549bed2d/nsis/lang/italian.nsi#L20C1-L33C1
```nsis
##############################################################################
# Translated license file for the license page                            {{{1
##############################################################################

LicenseLangString page_lic_file 0 "${VIMRT}\doc\uganda.nsis.txt"
#LicenseLangString page_lic_file ${LANG_ITALIAN} "${VIMRT}\doc\uganda.nsis.itx"

##############################################################################
# Translated README.txt file, which is opened after installation          {{{1
##############################################################################

LangString vim_readme_file 0 "README.txt"
#LangString vim_readme_file ${LANG_ITALIAN} "README.itx.txt"
```
.

I also plan to make “nsis/Makefile” and “nsis/Make_mvc.mak” files for creating the installer. But this will be done later.

**Comment by chrisbra** [comment]
Please suggest further enhancements. For now I have included @azc100 Italian README and License file.

**Comment by k-takata** [comment]
I think there is no need to append `x` to the extension of the `lang/{README,LICENSE}.ll` files because they are not Vim help files.
Simply `lang/{README,LICENSE}.ll` or `lang/{README,LICENSE}.ll.txt` might be better.

**Comment by azc100** [comment]
@k-takata 
I agree with you. I would go for simply for lang/{README,LICENSE}.ll

**Comment by chrisbra** [comment]
fixed by 2730d3873f222876cdd7459772fef6d935c23783
